sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -305,27 +327,49 @@ public void startBundle(StartBundleContext c) throws Exception {
     }
 
     @ProcessElement
+    @SuppressWarnings("ReferenceEquality")
     public void processElement(ProcessContext c) throws Exception {
-      List<OutgoingMessage> pubsubMessages = new ArrayList<>(publishBatchSize);
-      int bytes = 0;
+      // TODO(sjvanrossum): Refactor the write transform so this map can be indexed with topic +
+      // ordering key and have bundle scoped lifetime.
+      // NOTE: A single publish request may only write to one ordering key.
+      // See https://cloud.google.com/pubsub/docs/publisher#using-ordering-keys for details.
+      Map<String, OutgoingData> orderingKeyBatches = new HashMap<>();
+      @MonotonicNonNull String currentOrderingKey = null;
+      @Nullable OutgoingData currentBatch = null;
       for (OutgoingMessage message : c.element()) {
-        if (!pubsubMessages.isEmpty()
-            && bytes + message.getMessage().getData().size() > publishBatchBytes) {
-          // Break large (in bytes) batches into smaller.
-          // (We've already broken by batch size using the trigger below, though that may
-          // run slightly over the actual PUBLISH_BATCH_SIZE. We'll consider that ok since
-          // the hard limit from Pubsub is by bytes rather than number of messages.)
-          // BLOCKS until published.
-          publishBatch(pubsubMessages, bytes);
-          pubsubMessages.clear();
-          bytes = 0;
+        String messageOrderingKey = message.getMessage().getOrderingKey();
+        if (currentOrderingKey == null || !currentOrderingKey.equals(messageOrderingKey)) {
+          currentOrderingKey = messageOrderingKey;
+          currentBatch = orderingKeyBatches.get(currentOrderingKey);
+        }
+        if (currentBatch == null) {
+          currentBatch = new OutgoingData();
+          currentBatch.messages.add(message);
+          // TODO(sjvanrossum): https://github.com/apache/beam/issues/31800
+          currentBatch.bytes += message.getMessage().getData().size();
```

Comment:
Would it simplify this code if we remove these lines and drop the `else {` block below?

Reply:
Done.